Habemus Macam

jacg opened the pull request:
Closes #15.

The remaining problems against which we were banging our heads recently, were caused by linking multiple versions of libc++, which was caused by the dependencies (G4 and Qt) and the client code using different compiler configurations.

The solution[^1] is to use the gcc environment rather than the clang one in the flake's `devShell`. This is only necessary on Darwin, but (for now) it has been changed on Linux too.

The change to gcc breaks the compile-time tests we added recently (which have been removed, to make GHA pass).

Thus, before we can merge this, we have to:

+ [x] Fix failing compile-time tests
+ [x] Revert to clang on Linux

@soleti this has been tested on macOS on Intel, but we do not have access to Apple Silicon, so it would be good if you could check that it works there too.

(Why is the macOS GHA so slow? The one I've been waiting for before submitting this PR, completed on Linux in under 9 minutes, the macOS one chugged along for muuuch longer and then failed spuriously, so I had to restart it. Then it took another 21 minutes.)

[^1]: This solution is temporary: a fix is [in the pipeline](https://github.com/NixOS/nixpkgs/issues/234710) upstream.

soleti commented:
This looks like it's working on macOS Silicon as well, I am able to run `g4-examples/B1`.

jacg commented:
@soleti Could you also try the version in the `fix-clang-on-darwin` branch?

soleti commented:
Also works.

jacg commented:
This branch uses clang on Linux, and stdenv (g++) on Darwin. Over in `fix-clang-on-darwin` we have a version which uses clang on both, but forces the Darwin one to use the same version of libc++ as everything else. The latter is a bit ugly, and should become obsolete once the aforementioned [upstream fix](https://github.com/NixOS/nixpkgs/issues/234710) lands.

Which one do we prefer?

(I'll add the `fix-clang-on-darwin` solution here, for archival purposes: we can revert it if we want to stick with what is already here.)

`fix-clang-on-darwin` should sidestep the compile-time test failures on g++, but maybe we want to take this opportunity to address those. But I wouldn't delay merging this for the sake of fixing compile-time tests on g++.

g++ also spews an absurd amount of warnings.

There is also #17.

jacg commented:
> `fix-clang-on-darwin` should sidestep the compile-time test failures on g++

[Apparently](https://github.com/jacg/nain4/actions/runs/5888404966/job/15969543216), not.

gonzaponte commented:
> Which one do we prefer?

Will it still be ugly once the upstream fix is in place?

There is one advantage of keeping g++ on darwin and that's that we don't ignore g++, which we have been doing for a while. As you say:

> There is also https://github.com/jacg/nain4/issues/17.

> But I wouldn't delay merging this for the sake of fixing compile-time tests on g++.

When I get some time I will explore what's going on with that in a separate branch

> g++ also spews an absurd amount of warnings.

I guess this has to do with the different interpretations of `-Wshadow` and so on? If so we will fix it with pragmas like we did before.

jacg commented:
> Will it still be ugly once the upstream fix is in place?

Once the upstream fix is in place, it should look exactly like it did before we touched anything, but now it should work.

Currently the clang 16 env on Darwin uses a different version of libc++ from everything else, that's what was causing our problems. After the fix, it should use the same version. IIUC. So our old flake should Just Work.

> There is one advantage of keeping g++ on darwin and that's that we don't ignore g++, which we have been doing for a while. As you say:
> 
> > There is also #17.

I guess we should just add it to GHA.

Off the top of my head, it should be possible to have a default devShell with clang, and a second one with gcc. Then it should be possible to invoke the latter explicitly with something like `nix develop .#gcc`.

jacg commented:
> > `fix-clang-on-darwin` should sidestep the compile-time test failures on g++
> 
> [Apparently](https://github.com/jacg/nain4/actions/runs/5888404966/job/15969543216), not.

Fixed.

I've left the gcc-on-darwin-clang-on-linux solution in the history, for reference, and replaced it with the patch-clang-on-darwin-to-use-libc++-11 approach.

This avoids using g++, so the compile-time tests now pass.

I suggest that we merge, and deal with g++ related issues (broken compile-time tests, running g++ on GHA) in a separate PR.

gonzaponte commented on client_side_tests/client_fetch_content/CMakeLists.txt:
```diff
@@ -24,6 +24,14 @@ include(${Geant4_USE_FILE})
 
 add_executable(client_exe main.cc)
 
+if(APPLE)
+    target_link_options(Nain4 PRIVATE -undefined dynamic_lookup)
+endif()
+
+if(APPLE)
+    target_link_options(client_exe PRIVATE -undefined dynamic_lookup)
+endif()
```
why is Nain4 a target here, this is a client, shouldn't need to mess with Nain4's link options.

jacg replied:
I don't know *why* it's needed, but [here is the proof](https://github.com/jacg/nain4/actions/runs/5890431218/job/15975511647) *that* it is needed.

gonzaponte replied:
Ah, I think we go back again to the issue where fetchcontent grabs the last tag, not the current branch.
#67

jacg replied:
Ugh.

gonzaponte replied:
Can we have a commit changing `GIT_TAG` to your branch to convince ourselves that it's the only issue?

gonzaponte replied:
It does. It confirms what I was thinking. Now it fails in the recompilation stage because `GIT_TAG` is not a tag and therefore it attempts to use the network. I'm confident that the tests pass without the `Nain4` link option (as they should). We need to address #67 to avoid future problems like this.

gonzaponte replied:
So we revert back the `GIT_TAG` change and this will be ready to merge, I believe.

jacg replied:
So we tag and merge?

gonzaponte replied:
Ah wait, we need to update the documentation, right?

jacg replied:
Uhhh ... which part? I've got so many context switches going on right know that I have no idea whatsoever what I am doing any more.

gonzaponte replied:
https://jacg.github.io/nain4/how-to/enable-nain4-in-cmake.html

Although technically, dealing with stupid macos is not a nain4 problem. But I guess it's worth the trouble.

jacg replied:
So we'd have to add the if-apple-then-dynamic-lookup gubbins somewhere in there? 

Yes, that absolutely needs to go in.

Anything else?

jacg replied:
> dealing with stupid macos is not a nain4 problem

Unfortunately, in the harsh Real World, it is very much our problem.

gonzaponte replied:
I can't think of anything else. Just remember to revert [9d9d552](https://github.com/jacg/nain4/pull/82/commits/9d9d5526f00d216b1ee51cf6a2b92f907ced797d).

We can provide a CMake function for our clients to do if-apple-then-dynamic-lookup more conveniently, but that will probably not work out of the box and deserves its own PR.

gonzaponte replied:
The macos build is passing the tests. This means that the networkless recompilation test is not working as expected. My guess is that the command we use to restrict network usage is not working on macos. This means that it's allowed to use the network, fetch the branch and recompile, thus passing the test. But this is not supposed to happen.

gonzaponte replied:
And, [I can confirm that this is the issue](https://github.com/jacg/nain4/actions/runs/5890975701/job/15977141731?pr=82#step:8:121):

> sudo: iptables: command not found

gonzaponte commented on client_side_tests/client_subdirectory/CMakeLists.txt:
```diff
@@ -11,6 +11,14 @@ include(${Geant4_USE_FILE})
 
 add_executable(client_exe main.cc)
 
+if(APPLE)
+    target_link_options(Nain4 PRIVATE -undefined dynamic_lookup)
+endif()
+
+if(APPLE)
+    target_link_options(client_exe PRIVATE -undefined dynamic_lookup)
+endif()
+
```
same remark as above.

jacg replied:
This one does seem to be unnecessary.

gonzaponte commented on flake.nix:
```diff
@@ -22,29 +22,51 @@
+          just
+          gnused # For hacking CMAKE_EXPORT stuff into CMakeLists.txt
+          mdbook
+        ] ++ lib.optionals stdenv.isDarwin [
+
+        ] ++ lib.optionals stdenv.isLinux [
+        ];
```
This is not necessary anymore, right?

jacg replied:
Not necessary, but I want to keep it there, to show how it can be done.

jacg commented:
Do you want to review #86 and/or #84 before I merge them?

I'm bored of having to cancel a duplicate workflow with every push, so it would be good to get at least #86 out of the way quickly.

gonzaponte commented:
There seems to be a linking problem in macOS now?

gonzaponte commented:
> But it is needed

I see two options:
+ Now that we understand the failures, we remove the offending lines, merge this PR and tag the new master
+ We merge this PR with the offending lines, tag master at that point and push another commit removing the lines.

I don't have a strong preference.

jacg commented:
What do you consider to be the offending lines? I don't think there are any (unless you're referring to the apple dynamic links in the client-side tests, whose removal I have finally managed to push, so they are not there any more).

If I've got that right, then we

1. change the tag in the client-side test
2. merge
3. tag

done.

jacg commented:
You or me? I can do it, if you haven't started yet.

gonzaponte commented:
Go for it.
